Optimize Elasticsearch queries that fetch no columns

martint opened the pull request:
In that case, the connector only needs to produce pages with
a row count. We leverage Elasticsearch's count API.

martint commented on presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/client/ElasticsearchClient.java:
```diff
@@ -615,6 +607,45 @@ public SearchResponse nextPage(String scrollId)
+            try {
+                response = client.getLowLevelClient()
+                        .performRequest(
+                                "POST",
```
It's not surprising. It's probably not very strict about it. But yeah, it's supposed to be a GET according to the documentation. I'll adjust it.

aalbu reviewed (approved):
LGTM

findepi commented on presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/ElasticsearchPageSourceProvider.java:
```diff
@@ -53,10 +58,33 @@ public ConnectorPageSource createPageSource(
+
+            ImmutableList.Builder<Page> pages = ImmutableList.builder();
+            while (count > 0) {
+                int batch = Ints.saturatedCast(count);
+                pages.add(new Page(batch));
```
@sopel39 is it safe to inject 2 billion row pages into the system?

in case of `select count(*) from ..`, they are going to be partial aggregated quickly
what about `select 1 from ..` query?

(Martin, if you make pages, smaller, make it also lazy, as `pages` list  may grow big. Currently, it's well consttrained by my imagination)

martint replied:
These pages are empty, actually. The only thing they carry is a count.

findepi replied:
Yes, I know. But if you project a column (random or constant), they become non-empty and _i think_ ScanFilterProject won't be able to split this into multiple small ones. Will it?

findepi commented on presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/ElasticsearchPageSourceProvider.java:
```diff
@@ -53,10 +58,33 @@ public ConnectorPageSource createPageSource(
+            long count = client.count(
+                    elasticsearchSplit.getIndex(),
+                    elasticsearchSplit.getShard(),
+                    buildSearchQuery(session, elasticsearchTable.getConstraint().transform(ElasticsearchColumnHandle.class::cast), elasticsearchTable.getQuery()));
```
`ScanQueryPageSource` is going to repeat the same logic.
Maybe do it here once, and pass to `client.count` or `ScanQueryPageSource`?

martint replied:
That's less ideal. I consider that an implementation detail of the specific page source. This branch of the `if` is really a "CountQueryPageSource", but I thought it overkill to create a dedicated class for it.

I moved this whole block into a separate dedicated method.

findepi commented on presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/client/ElasticsearchClient.java:
```diff
@@ -613,6 +605,45 @@ public SearchResponse nextPage(String scrollId)
+                throw new PrestoException(ELASTICSEARCH_INVALID_RESPONSE, e);
+            }
+        }
+        finally {
+            countStats.add(Duration.nanosSince(start));
+        }
```
would be nice to use sth like ThriftMetastoreApiStats -- less verbose and count failures too

(no change requested)

findepi commented on presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/client/ElasticsearchClient.java:
```diff
@@ -663,6 +701,31 @@ public TimeStat getNextPageStats()
         return handler.process(body);
     }
 
+    private static PrestoException propagate(ResponseException exception)
```
would be nice to extract this in a separate "extract .." commit

(i assume nothing material changed beyond that btw)

findepi commented on presto-elasticsearch/src/test/java/io/prestosql/elasticsearch/BaseElasticsearchSmokeTest.java:
```diff
@@ -739,6 +740,15 @@ public void testMultiIndexAlias()
+        assertQuery("SELECT count(*) FROM orders");
+        assertQuery("SELECT count(*) FROM orders WHERE orderkey > 10");
+        assertQuery("SELECT count(*) FROM (SELECT * FROM orders LIMIT 10)");
+        assertQuery("SELECT count(*) FROM (SELECT * FROM orders WHERE orderkey > 10 LIMIT 10)");
```
as we are heading towards more agg pushdowns, you can add this as generic test cases in 
`io.prestosql.testing.AbstractTestIntegrationSmokeTest#testCountAll`

martint committed:
Extract error handling processing to a separate method

martint committed:
Optimize Elasticsearch queries that fetch no columns

In that case, the connector only needs to produce pages with
a row count. We leverage Elasticsearch's count API.

martint commented:
@findepi, updated

findepi commented on presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/CountQueryPageSource.java:
```diff
@@ -0,0 +1,90 @@
+class CountQueryPageSource
+        implements ConnectorPageSource
+{
+    private static final int BATCH_SIZE = 10000;
```
please capture some motivation how we ended up with a modest number as a comment

findepi commented on presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/CountQueryPageSource.java:
```diff
@@ -0,0 +1,90 @@
+{
+    private static final int BATCH_SIZE = 10000;
+
+    private final long readTime;
```
readTime -> readTimeNanos

(btw we have `readTimeNanos` in `ScanQueryPageSource`)
